untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+
+_Application_ is the operation of substituting a specific value for the parameter in the abstraction.
+It is the only operation defined in the pure untyped lambda calculus.
+In the standard notation, the application is denoted simply by whitespace.
+For example, $f\; x$ is the application of $f$ to $x$.
+Since all abstractions are univariate by definition, there isn't a special syntax for multivariate application.
+
+You might notice that using whitespace can lead to ambiguity.
+We'll use parentheses where necessary to avoid it.
+Additionally, we will assume that application is left-associative and binds more tightly than abstractions.
+Hence, $f\;a\;b\;c$ is $((f\;a)\;b)\;c$, i.e. $f$ is applied to $a$, then the result is applied to $b$, then to $c$.
+Since application binds more tightly, we don't need parentheses around the body of an abstraction.
+However, we might need parenthesis around the whole abstraction itself.
+We will also eschew parentheses for nested abstractions, e.g. $\lambda x. \lambda y. x\;y$ is $\lambda x. (\lambda y. (x\;y)).$
+
+Variables in the body of an abstraction that are parameters of said abstraction are called _bound variables_.
+Other variables are called _free variables_.
+
+An expression without free variables is called a _closed term_ or a _combinator_.
+
+The choice of names of bound variables is arbitrary.
+Expressions that differ only in the names of bound variables are called _alpha-equivalent_.
+For example, $\lambda x. x$ and $\lambda y. y$ are alpha-equivalent, but $\lambda x. y$ and $\lambda y. z$ are not.
+
+## Defining the computation
+
+The step of the computation is called beta-reduction.
+It is a single application of one expression to another.
+Formally, beta-reduction uses the following rules.
+Given an expression of the form $(\lambda x. t_1)\; t_2$ (i.e. an application of some abstraction to some expression $t_2$):
+
+1. Rename bound variables in $t_1$ and $t_2$ to avoid ambiguity.
+    The new expressions, $t_1'$ and $t_2',$ must be alpha-equivalent to $t_1$ and $t_2,$ respectively.
+2. Replace all the instances of the bound variable $x$ in $t_1'$ with $t_2'.$
+    That is the result of the computation.
+
+Any expression where beta-reduction can be applied, we'll call a redex (**red**ucible **ex**pression).
+
+For example, the expression $(\lambda x. x) y$ is a redex, and one step of beta-reduction transforms it to just $y$.
+
+Per rule 1, we rename bound variables if there might be ambiguity.
+For example, $(\lambda x. \lambda y. x\;y)\;(\lambda x. x\;y)$ is a redex.
+However, the $y$ identifier is ambiguous.
+In the left term, it refers to the bound variable; in the right one, it refers to a free variable.
+To avoid ambiguity, we can rename $y$ in the first expression to, say, $z$: $(\lambda x. \lambda z. x\;z)\;(\lambda x. x\;y).$
+Then we can do the rest of the steps with no fear of confusion: $\lambda z. (\lambda x. x\;y)\;z.$
+
```

Comment:
Feel like this should be a subsection called evaluation strategies.

Reply:
@NaeosPsy I've added a level 3 header where it seemed appropriate (i.e. after "It brings us to the discussion of evaluation strategies."). The highlighted part is not it, though.

@Heimdell :unamused: I actually introduce de Bruijn encoding in the exercises section. It's nice when implementing λ-calculus in software, but not so well suited for napkin math we're doing in this here article.

BTW, you might want to check my explanation of de Bruijn encoding, I'm not sure if it's quite on-point because it's been a while since I implemented my λ-calculus evaluation engine.